exonum-java-binding/core/rust/src/runtime/service_factory_adapter.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use exonum::{
+    blockchain::Service,
+    helpers::fabric::{Command, CommandExtension, Context, ServiceFactory},
+};
+use runtime::{cmd::Run, config::Config, java_service_runtime::JavaServiceRuntime};
+use std::sync::{Once, ONCE_INIT};
+
+static mut JAVA_SERVICE_RUNTIME: Option<JavaServiceRuntime> = None;
+static JAVA_SERVICE_RUNTIME_INIT: Once = ONCE_INIT;
+static JAVA_SERVICE_CREATE_COMMAND: Once = ONCE_INIT;
+
+/// Adapts current single-service interface of Exonum to multiple EJB services until dynamic
+/// services are implemented. Tracks the `JavaServiceRuntime` instantiation and command extension
+/// and represents factory for every user service relying on current `ServiceFactory` interface.
+pub struct JavaServiceFactoryAdapter {
+    name: String,
+    artifact_path: String,
+}
+
+impl JavaServiceFactoryAdapter {
+    /// Creates new instance with given service name and path to the artifact.
+    pub fn new<S: AsRef<str>>(name: S, artifact_path: S) -> Self {
+        JavaServiceFactoryAdapter {
+            name: name.as_ref().to_owned(),
+            artifact_path: artifact_path.as_ref().to_owned(),
+        }
+    }
+
+    // Creates new runtime from provided config or returns the one created earlier. There can be
+    // only one `JavaServiceRuntime` instance at a time.
+    fn get_or_create_java_service_runtime(context: &Context) -> JavaServiceRuntime {
+        // Initialize runtime if it wasn't created before.
+        JAVA_SERVICE_RUNTIME_INIT.call_once(|| {
+            // Prepare config for service runtime configuration
+            let config = Self::extract_config(context);
+
+            let runtime = JavaServiceRuntime::create_java_runtime(config);
+            unsafe {
+                JAVA_SERVICE_RUNTIME = Some(runtime);
+            }
+        });
+
+        unsafe { JAVA_SERVICE_RUNTIME.clone() }
+            .expect("Trying to return runtime, but it's uninitialized")
+    }
+
+    // Extracts EJB and JVM configuration from Context
+    fn extract_config(context: &Context) -> Config {
```

Comment:
Why don't we push this operation to the ServiceFactory implementation and pass a Config to the method above so that these operations on the Context are closer to where they make sense (ServiceFactory)?

Reply:
Well, the main idea was to extract Config from Context only once and when it's only required - at the service runtime creation step. In other words - why any other service that is being created after the creation of the service runtime should depend on possible errors of extracting Config from Context (even if they aren't logically possible within current implementation) if Context is not needed anymore? And WRT to the assumption that this code is temporary until dynamic services in core I considered this acceptable.

Reply:
Yes, it is acceptable and may remain as is. 

Is this context mutable for that hypothetical failure to be a concern?

I see that this change will involve extracting the config even when it is not needed — but this is neither a frequent nor expensive operation to be a concern. The temporary part here is likely ServiceFactory and Context, not the runtime. So it also makes sense to move this operation operating on the context to the factory, and pass to JSR method a thing that it understands — its configs.

Reply:
Yes, that sounds reasonable.

Reply:
Corrected.